test/integration/apiserver/crd_validation_expressions_test.go

```diff
@@ -0,0 +1,470 @@
+		structuralWithValidators := crdWithSchema(t, "InvalidStructuralMetadata", structuralSchemaWithInvalidMetadataValidators)
+		_, err := fixtures.CreateNewV1CustomResourceDefinition(structuralWithValidators, apiExtensionClient, dynamicClient)
+		if err == nil {
+			t.Error("Expected error creating custom resource but got none")
+		} else if !strings.Contains(err.Error(), "undefined field 'labels'") {
+			t.Errorf("Expected error to contain %s but got %v", "undefined field 'labels'", err.Error())
+		}
+	})
+	t.Run("CRD creation MUST pass if a x-kubernetes-validations rule accesses metadata.name", func(t *testing.T) {
+		structuralWithValidators := crdWithSchema(t, "ValidStructuralMetadata", structuralSchemaWithValidMetadataValidators)
+		_, err := fixtures.CreateNewV1CustomResourceDefinition(structuralWithValidators, apiExtensionClient, dynamicClient)
+		if err != nil {
+			t.Error("Unexpected error creating custom resource but metadata validation rule")
+		}
+	})
+}
+
+func crdWithSchema(t *testing.T, kind string, schemaJson []byte) *apiextensionsv1.CustomResourceDefinition {
+	plural := strings.ToLower(kind) + "s"
+	crd := fixtures.NewNoxuV1CustomResourceDefinition(apiextensionsv1.ClusterScoped)
+	var c apiextensionsv1.CustomResourceValidation
+	err := json.Unmarshal(schemaJson, &c)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("%s.mygroup.example.com", plural)},
+		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+			Group: "mygroup.example.com",
+			Versions: []apiextensionsv1.CustomResourceDefinitionVersion{{
+				Name:    "v1beta1",
+				Served:  true,
+				Storage: true,
+				Schema:  &c,
+				Subresources: &apiextensionsv1.CustomResourceSubresources{
+					Status: &apiextensionsv1.CustomResourceSubresourceStatus{},
+				},
+			}},
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Plural: plural,
+				Kind:   kind,
+			},
+			Scope: apiextensionsv1.ClusterScoped,
+		},
+	}
+	return crd
+}
+
+var structuralSchemaWithValidators = []byte(`
+{
+  "openAPIV3Schema": {
+    "description": "CRD with CEL validators",
+    "type": "object",
+	"x-kubernetes-validations": [
+	  {
+		"rule": "spec.x + spec.y >= (has(self.status) ? self.status.z : 0)"
+	  }
+	],
+    "properties": {
+      "spec": {
+        "type": "object",
+        "properties": {
+          "x": {
+            "type": "integer",
+			"default": 0
+          },
+          "y": {
+            "type": "integer",
+			"default": 0
+          },
+          "extra": {
+			"type": "string",
+			"x-kubernetes-validations": [
+			  {
+				"rule": "self.startsWith('anything')"
+			  }
+			]
+          },
+		  "floatMap": {
+			"type": "object",
+			"additionalProperties": { "type": "number" },
+			"x-kubernetes-validations": [
+			  {
+				"rule": "self.all(k, self[k] >= 0.2)"
```

Comment:
CEL doesn't support maps as ranges to comprehensions in the spec, but it's implemented in Go. This is an area for improvement as well since you only get the key and not the value as part of the bounded iteration and there's no way to reconstruct a map in `map` and `filter` macros.

Reply:
Since anything we expose to the Kubernetes community follows Hyrum's Law. What should we do? Do we want the functionality even though it's not in the spec? Would it be possible for us to introduce a comprehension for iterating over map elements in the near term? We're going to have to lock down our usage of CEL for Kubernetes pretty quick for compatibility reasons.